xds/src/main/java/io/grpc/xds/internal/security/trust/XdsX509TrustManager.java

```diff
@@ -303,4 +306,44 @@ public X509Certificate[] getAcceptedIssuers() {
     }
     return delegate.getAcceptedIssuers();
   }
+
+  public static boolean verifyDnsNameWildcard(
+          String altNameFromCert, String sanToVerify, boolean ignoreCase) {
+    String[] splitPattern = splitAtFirstDelimiter(ignoreCase
+        ? sanToVerify.toLowerCase(Locale.ROOT) : sanToVerify);
+    String[] splitDnsName = splitAtFirstDelimiter(ignoreCase
+        ? altNameFromCert.toLowerCase(Locale.ROOT) : altNameFromCert);
+    if (splitPattern.length < 2 || splitDnsName.length < 2) {
+      return false;
+    }
+    if (splitPattern[0].contains("*")
+        && !splitPattern[1].contains("*")
+        && !splitPattern[0].startsWith("xn--")) {
+      return splitDnsName[1].equals(splitPattern[1])
+          && labelWildcardMatch(splitDnsName[0], splitPattern[0]);
+    }
+    return false;
+  }
+
+  private static boolean labelWildcardMatch(String dnsLabel, String pattern) {
+    final char glob = '*';
+    if (pattern.length() == 1 && pattern.charAt(0) == glob) {
+      return true;
+    }
+    if (pattern.chars().filter(ch -> ch == glob).count() == 1) {
```

Comment:
in terms of performance it can be
```java
int globIndex = pattern.indexOf(glob);
if (pattern.indexOf(glob, globIndex + 1) == -1) {
  return dnsLabel.length() >= pattern.length() - 1
      && dnsLabel.startsWith(pattern.substring(0, globIndex))
      && dnsLabel.endsWith(pattern.substring(globIndex + 1));
```